### PR TITLE
  livebook: Use `mix release` to build instead of escript

### PR DESCRIPTION
**Pull request by Munksgaard**
## Description of changes

The current build of livebook does not work with the new [Livebook Teams](https://livebook.dev/teams/) features. The problem can be observed by running the current version of livebook, adding a new team and going to the team page. The process will crash and the team page will show a 500 error.

The base of the problem is that the escript build method is not officially supported. This commit changes the livebook package to use the `mix release` workflow, which is also the one used to build the official Docker container.

Unfortunately, the binary built with `mix release` does not support command line arguments like the `escript` binary does. Instead, users need to pass in most of the configuration as environment variables, as documented [here](https://hexdocs.pm/livebook/readme.html#environment-variables). As a result, this commit also changes the Livebook service to reflect this new way of configuring Livebook.

Finally, the Livebook release configuration specifically excludes the ERTS (Erlang Runtime System), which means that the resulting release cannot run without Erlang installed.

I have tested the results (both of the package and the service) locally.

cc. @NixOS/beam 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/development/livebook.nix [hunk 1]
```diff
@@ -80,18 +65,11 @@ in
...
-          let
-            args = lib.cli.toGNUCommandLineShell { } ({
-              inherit (cfg) port;
-              ip = cfg.address;
-              name = cfg.erlang_node_name;
-              sname = cfg.erlang_node_short_name;
-            } // cfg.options);
-          in
-            "${cfg.package}/bin/livebook server ${args}";
+        EnvironmentFile = optional (cfg.environment != null) cfg.environmentFile;
```
Comment (happysalada) [inline nixos/modules/services/development/livebook.nix:68]: shouldn't it be cfg.environmentFile != null ?

also I've never seen the environment attribute before, will it merge correctly the environmentFile and the environment attribute set ?

Comment (Munksgaard) [reply nixos/modules/services/development/livebook.nix:68]: > shouldn't it be cfg.environmentFile != null ?

Yes! Will fix right away.

> also I've never seen the environment attribute before, will it merge correctly the environmentFile and the environment attribute set ?

It is used in other NixOS service definitions, such as [oci-containers](https://search.nixos.org/options?channel=unstable&show=virtualisation.oci-containers.containers.%3Cname%3E.environment&from=0&size=50&sort=relevance&type=packages&query=virtualisation.oci-containers.containers.%3Cname%3E.environment), but I am happy to give it another name if desired. 

Why both `environment` and `environmentFile`? I think it makes sense to be able to divide up sensitive information, such as `LIVEBOOK_PASSWORD` and put that in `environmentFile` and non-sensitive information, such as `LIVEBOOK_PORT` and put that in `environment.

The relationship between the two arguments are defined [here](https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#EnvironmentFile=). It is my understanding that `EnvironmentFile` supersedes `Environment`.

Comment (happysalada) [reply nixos/modules/services/development/livebook.nix]: I think the separation makes sense.

I thought it was environment with a capital e, but it's not. I've even used this attribute before in one of the modules I made, never mind.

Comment (scvalex) [reply nixos/modules/services/development/livebook.nix]: I'd go further and say we shouldn't have both `environment` and `environmentFile`.

The main reason to use `environmentFile` over `environment` is that you can avoid putting your password in the nix store.  But if you don't care about this, you can easily just generate a file with `pkgs.writeText` like the test does.  So, I'd propose we get rid of `environement` and give an example of using `pkgs.writeText` in the docs.

**Comment by happysalada** [comment]
This looks good to me, leaving this open for a bit to give others the chance to have a look.

File: nixos/modules/services/development/livebook.nix (hunk 1 above)
Comment (Munksgaard) [reply nixos/modules/services/development/livebook.nix]: > I'd go further and say we shouldn't have both `environment` and `environmentFile`.

I don't agree, cf. [my comment above](https://github.com/NixOS/nixpkgs/pull/283447#discussion_r1464803951). I think they can serve different purposes at the same time.

Comment (happysalada) [reply nixos/modules/services/development/livebook.nix]: I don't have a strong opinion either way on this, I'm happy for both of you to reach the compromise that works best for everyone.

**Comment by scvalex** [comment]
GitHub ate my comment on the change, so I'll post it here.

> I think the separation [between environment and environmentFile] makes sense.

Does it, though?  I think I'd be pretty confused as a user if I saw both of these options.  It would not at all be obvious to me that you're supposed to use both, one for secrets and one for other options.  I'd also be wondering how the two options merge.

How about not having the `environment` option at all and instead documenting the use of `pkgs.writeText` to inline the options in the nix file?  We can then add the appropriate warning about storing secrets in the nix store in the docs.

**Comment by Munksgaard** [comment]
> GitHub ate my comment on the change, so I'll post it here.
> 
> > I think the separation [between environment and environmentFile] makes sense.
> 
> Does it, though? I think I'd be pretty confused as a user if I saw both of these options. It would not at all be obvious to me that you're supposed to use both, one for secrets and one for other options. I'd also be wondering how the two options merge.
> 
> How about not having the `environment` option at all and instead documenting the use of `pkgs.writeText` to inline the options in the nix file? We can then add the appropriate warning about storing secrets in the nix store in the docs.

Perhaps some of that can be alleviated using documentation?

Also, I will note that there is [plenty](https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=virtualisation.oci-containers.containers.%3Cname%3E.env) [of](https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=services.vaultwarden.environmen) [precedence](https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=services.lldap.env) [for](https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=services.garage) [this](https://search.nixos.org/options?channel=unstable&from=0&size=50&sort=relevance&type=packages&query=services.mastodon.extra).

**Comment by scvalex** [comment]
Actually, yeah!  Those are  good examples. Compare [`oci-containers`](https://search.nixos.org/options?channel=unstable&show=virtualisation.oci-containers.containers.%3Cname%3E.environmentFiles&from=0&size=50&sort=relevance&type=packages&query=virtualisation.oci-containers.containers.%3Cname%3E.env) with [`vaultwarden`](https://search.nixos.org/options?channel=unstable&show=services.vaultwarden.environmentFile&from=0&size=50&sort=relevance&type=packages&query=services.vaultwarden.environmen).

The former looks pretty bad: you've got two very similar sounding options and, if you're a user, it's not at all clear what you should configure where.  The latter documents what the point of `environmentFile` is.  I'd be fine with expanding the options documentation like this.  We actually have a comment to this extent in the manual, but there's no link from the options to the manual.

So, if we don't want to get rid of `environment`, I'd say we should at least copy some of the manual text into one of the two options, and also clarify in the warning in the manual that the two options have different security implications.

**Comment by Munksgaard** [comment]
Great suggestions, I'll do that later today :+1:

**Comment by scvalex** [comment]
Also, I'm trying to run this locally, but I'm seeing errors when pulling in dependencies (i.e. when evaluating a notebook with dependencies that need to be downloaded and compiled):

```
$ LIVEBOOK_PORT=8082 LIVEBOOK_IFRAME_PORT=8083 result/bin/livebook start
```

image

We used to add Elixir to the PATH through `wrapProgram` and we also added some Rebar3 path.  It looks like we need to add those to the service now.

**Comment by Munksgaard** [comment]
> We used to add Elixir to the PATH through `wrapProgram` and we also added some Rebar3 path. It looks like we need to add those to the service now.

Hm, that shouldn't be necessary now that I've included the erts, but I'll take a look at it. Thanks for trying it out.

**Comment by scvalex** [comment]
The Livebook dependency mechanism is built on top of `Mix.install` and that wouldn't be included in erts.

**Comment by Munksgaard** [comment]
You're right, I can reproduce on my own machine. Will fix.

**Comment by Munksgaard** [comment]
> We used to add Elixir to the PATH through `wrapProgram` and we also added some Rebar3 path. It looks like we need to add those to the service now.

I don't understand why Rebar3 is needed. Is it just an extra precaution, or is there something I am not aware of?

**Comment by scvalex** [comment]
Works for me too.  I tried some of the Learn notebooks and also a few of my own.

I can't try the Teams feature because I'm not in the beta.

**Comment by scvalex** [comment]
I also need the "Livebook Teams key" that you have.

**Comment by scvalex** [comment]
Alright, the teams stuff works for me too.

**Comment by Munksgaard** [comment]
> are you both satisfied with the module ?
> (personally I am, I just want to make sure we are good to go).

No, I still have some documentation fixes to make. Additionally, the original issue that was caused by using escript has actually been [fixed](https://github.com/livebook-dev/livebook/pull/2439), so the teams functionality works even without this PR, at least once there's a new release of Livebook. I am unsure if we'd rather stick to the old way of doing it (accepting that the teams feature does not work until there's a new release), or merge this PR and change how livebook is built.

**Comment by scvalex** [comment]
I don't have a strong preference, but I do like the `mix release` process from this PR a bit more because it makes some functionality like remote IEx shells available out of the box.

**Comment by Munksgaard** [comment]
> I don't have a strong preference, but I do like the `mix release` process from this PR a bit more because it makes some functionality like remote IEx shells available out of the box.

I've stuck with the `mix release` version and updated the documentation now. Please take a look at your convenience :-)

**Comment by scvalex** [comment]
Awesome!  Looks good to me.

**Comment by Munksgaard** [comment]
@happysalada I think we are ready to merge :champagne:

**Comment by happysalada** [comment]
@Munksgaard can you rebase to fix the conflict?

**Comment by happysalada** [comment]
Thank you both for your efforts on this!

**Pull request merged**